components/tidb_query/src/codec/mysql/decimal.rs

```diff
@@ -2251,22 +2242,10 @@ pub trait DecimalDecoder: NumberDecoder {
 
     /// `read_decimal_from_chunk` decode Decimal encoded by `write_decimal_to_chunk`.
     fn read_decimal_from_chunk(&mut self) -> Result<Decimal> {
-        let buf = self.bytes();
-        if buf.len() <= 4 {
-            return Err(Error::unexpected_eof());
-        }
-        let int_cnt = buf[0];
-        let frac_cnt = buf[1];
-        let result_frac_cnt = buf[2];
-        let negative = buf[3] == 1;
-        self.advance(4);
-
-        let mut d = Decimal::new(int_cnt, frac_cnt, negative);
-        d.result_frac_cnt = result_frac_cnt;
-
-        for id in 0..WORD_BUF_LEN {
-            d.word_buf[id as usize] = self.read_i32_le()? as u32;
-        }
+        let buf = self.read_bytes(DECIMAL_STRUCT_SIZE)?;
+        let d = Decimal::default();
+        let dst = &d as *const Decimal as *mut u8;
+        unsafe { copy_nonoverlapping(buf.as_ptr(), dst, DECIMAL_STRUCT_SIZE) }
```

Comment:
Prefer https://doc.rust-lang.org/std/mem/fn.transmute_copy.html to eliminate some unnecessary copy

Reply:
`transmute_copy` requires to retain a reference(not pointer) to src, while we can use `std::ptr::read_unaligned` instead:
```rust
let src = mem::transmute::<*const u8, *const Decimal>(buf.as_ptr());
ptr::read_unaligned(src)
```

Reply:
I got warning below using `as *const Decimal`
```shell
error: casting from `*const u8` to a more-strictly-aligned pointer (`*const codec::mysql::decimal::Decimal`) (1 < 4 bytes)
    --> components/tidb_query/src/codec/mysql/decimal.rs:2247:17
     |
2247 |         let p = buf.as_ptr() as *const Decimal;
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: `#[deny(clippy::cast_ptr_alignment)]` on by default
```

Is there a better way to do this?